### PR TITLE
Scale

### PR DESCRIPTION
**Pull request by ClashLuke**
Observation:
1) Adam gets unstable at scale, as gradients become smaller and sparser. This causes beta2 to decay to 0. However, once one gradient spike happens, beta1 will shoot up almost immediately, causing a large number to get divided by `eps` which disturbs the gradients a lot. Current solution: SM3. Looking into MADGRAD and Mirror-MADGRAD to fix this. Untested if Shampoo is affected as well.
2) Distributed Dataloaders are difficult but possible. Current solution: Only one Dataloader and no input stacking.
3) Small-Init + input norm causes significant instability (of the bad kind, see Figure 1). Solution: Revert to normal init.
4) FP64 in Jax works out the box, but there were missing casts in Shampoo. Now running the entire optimizer in FP64.
5) Dying ReLU is a real problem. Many nodes simply die (or get their weights down to 0) and they will rarely be activated again. Also, LeakyReLU performs worse than Swish/GeLU/Mish/LiSHT. Switching to LeCunTanH which doesn't have these problems, still allows for inversion and gives similar or better results on toy datasets. New Problem: Pruning won't be possible.

Figure 1:
grafik

### CODE REVIEW

**Comment by ClashLuke** [comment]
Shampoo improves stability:
grafik
grafik

Keeping it.

**Comment by ClashLuke** [comment]
This was probably the most difficult thing to find:
grafik

We used different inputs on each model-parallel device in all previous attempts (with the sharded Dataloader).

**Comment by ClashLuke** [comment]
Convergence works on v3-32s. Current training model with 16 blocks (2.5B) on one preemptible v3-32.
ToDo:
- [ ] Mid-Training Checkpoint+Resume
- [ ] Test MADGRAD (might solve Adam instability without diagonal SM3 approximation)
- [ ] Make DeepSource happy

**Comment by ClashLuke** [comment]
It still diverges. Trying alternatives to gradient descent now
grafik

**Comment by ClashLuke** [comment]
HeavyBall-Adam improves convergence a good bit compared to pure Adam (same speed):
grafik
grafik

Nesterov-HeavyBall-Adam improves convergence further (same speed) and allows grafting onto it (as beta2 will never be 0 -> step size won't go to infinity)
grafik
grafik

**Comment by ClashLuke** [comment]
Reducing batch size and LR decreases final accuracy
grafik
grafik

**Comment by ClashLuke** [comment]
Increasing depth helps with convergence per step
grafik
While decreasing convergence per time
grafik


ReZero helps in the early training (unknown if it'll stay the same during the late stages)
grafik

Maximizing the context size (with the same number of tokens) increases batch-to-batch variance, likely because it's seeing so many different datasets.
grafik
Potential solution: Train short for a few epochs and "fine-tune" a longer context onto the model, as in [BERT](https://arxiv.org/abs/1810.04805) and [Shortformer](https://arxiv.org/abs/2012.15832)

**Comment by ClashLuke** [comment]
Similarly, with a depth of 16 (above has a depth of 1), ReZero (purple) outperforms no output normalization (blue) but doesn't converge as well as SandwichLN (red):
grafik

**Comment by ClashLuke** [comment]
for some reason, increasing the width of the network still hurts convergence
grafik

**Comment by ClashLuke** [comment]
Losses finally match
grafik

**Comment by ClashLuke** [comment]
Model finally converges and v3-32 seems marginally better than v3-8:
grafik
Not trying to match the new v3-8 loss with the old loss.

**Comment by ClashLuke** [comment]
The main difference likely is the 0-init in the output embed which helps improve stability
grafik

**Comment by ClashLuke** [comment]
The core problem seems to be that it doesnt converge faster with more devices

**Comment by ClashLuke** [comment]
New scan+psum_scatter+all_gather-based gradients are faster and more accurate, yielding the same (or better) loss curve as with the non-parallel baseline loss:
grafik

Now onto fixing the difference between the old (blue) and new (orange/grey) code.

**Comment by ClashLuke** [comment]
Together with LeakyReLU we're back at the old performance:
grafik

Testing Mish now.

**Comment by ClashLuke** [comment]
Works, but v3-32 still doesn't work as well as v3-8
grafik

**Comment by ClashLuke** [comment]
LeakyReLU (0.01) > Mish > LecunTanh, and even the bumps are still matched
grafik

**Comment by ClashLuke** [comment]
I looked over the differences once again and solved all deepsource issues except for missing docstrings. Additionally, this branch introduces some pytests to ensure the custom gradients are actually correct this time (they weren't before).\
Merging, as the state is stable and significantly better than the current `main`.\
Convergence with v3-32s needs to be addressed in a separate branch.
